zcash_client_sqlite/src/wallet/scanning.rs

```diff
@@ -0,0 +1,1089 @@
+    RightFirstDisjoint,
+}
+
+impl RangeOrdering {
+    fn cmp<A: Ord>(a: &Range<A>, b: &Range<A>) -> Self {
+        use Ordering::*;
+        assert!(a.start <= a.end && b.start <= b.end);
+        match (a.start.cmp(&b.start), a.end.cmp(&b.end)) {
+            _ if a.end <= b.start => RangeOrdering::LeftFirstDisjoint,
+            _ if b.end <= a.start => RangeOrdering::RightFirstDisjoint,
+            (Less, Less) => RangeOrdering::LeftFirstOverlap,
+            (Equal, Less) | (Greater, Less) | (Greater, Equal) => RangeOrdering::LeftContained,
+            (Equal, Equal) => RangeOrdering::Equal,
+            (Equal, Greater) | (Less, Greater) | (Less, Equal) => RangeOrdering::RightContained,
+            (Greater, Greater) => RangeOrdering::RightFirstOverlap,
+        }
+    }
+}
+
+enum Joined {
+    One(ScanRange),
+    Two(ScanRange, ScanRange),
+    Three(ScanRange, ScanRange, ScanRange),
+}
+
+fn join_nonoverlapping(left: ScanRange, right: ScanRange) -> Joined {
+    assert!(left.block_range().end <= right.block_range().start);
+
+    if left.block_range().end == right.block_range().start {
+        if left.priority() == right.priority() {
+            Joined::One(ScanRange::from_parts(
+                left.block_range().start..right.block_range().end,
+                left.priority(),
+            ))
+        } else {
+            Joined::Two(left, right)
+        }
+    } else {
+        // there is a gap that will need to be filled
+        let gap = ScanRange::from_parts(
+            left.block_range().end..right.block_range().start,
+            ScanPriority::Historic,
+        );
+
+        match join_nonoverlapping(left, gap) {
+            Joined::One(merged) => join_nonoverlapping(merged, right),
+            Joined::Two(left, gap) => match join_nonoverlapping(gap, right) {
+                Joined::One(merged) => Joined::Two(left, merged),
+                Joined::Two(gap, right) => Joined::Three(left, gap, right),
+                _ => unreachable!(),
+            },
+            _ => unreachable!(),
+        }
+    }
+}
+
+fn insert(current: ScanRange, to_insert: ScanRange) -> Joined {
+    fn join_overlapping(left: ScanRange, right: ScanRange, insert: Insert) -> Joined {
+        assert!(
+            left.block_range().start <= right.block_range().start
+                && left.block_range().end > right.block_range().start
+        );
+
+        // recompute the range dominance based upon the queue entry priorities
+        let dominance = match insert {
+            Insert::Left => dominance(&right.priority(), &left.priority(), insert),
+            Insert::Right => dominance(&left.priority(), &right.priority(), insert),
+        };
+
+        match dominance {
+            Dominance::Left => {
+                if let Some(right) = right.truncate_start(left.block_range().end) {
+                    Joined::Two(left, right)
+                } else {
+                    Joined::One(left)
+                }
```

Comment:
If `right` (outside the `if let`) is empty then `join_overlapping` should return `Joined::One(left)`, so I think it's the correct behaviour for `truncate_start` to return `None` in that case. I will check the other uses of `truncate_start` and `truncate_end` to make sure.